include/zephyr/net/wifi.h

```diff
@@ -24,4 +26,10 @@ enum wifi_security_type {
 #define WIFI_CHANNEL_MAX 14
 #define WIFI_CHANNEL_ANY 255
 
+struct wifi_api {
+	struct ethernet_api eth_api;
+
+	/* Any WiFi specific functionality can be added here. */
+};
+
```

Comment:
This pulls in ethernet_api structure to all drivers, even tough most of them (all besides ESP32 as of now) do not (explicitly, at least to Zephyr knowledge) operate on ethernet frames at all.

Reply:
@mniestroj My initial PR was using ifdefs but I get @rlubos idea. Any suggestion regarding this so I can update this?

Reply:
Hi, any comments?

Reply:
@sylvioalves Can you elaborate on this topic a bit?

I recall that when we initially discussed native WiFi support in Zephyr it was said that WiFi is compatible with Ethernet at L2 level? That's why we decided to reuse the existing Ethernet L2, instead of implementing a separate L2 for WiFi, which would be way more complex task. I'm not expert in the area, but for sure this approach works here for Espressiff as well as for Nordic devices, where we use the same approach of reusing Ethernet L2.

Reply:
@mniestroj @rlubos It works with Ethernet L2. For now, there is no plan of changing this. I do agree we should add in Zephyr's roadmap `WiFi native` support, which means among other implementations, adding wpa_supplicant support.  
While we don't have it, we can proceed choosing the options discussed before:
1) Use `#ifdef` to allow Ethernet L2 + WiFi struct, which won't allow simultaneous `wifi offloaded` and `non-offloaded` usage, which is pretty much fine by me.
2) Keep as is and increase resource usage of `wifi offloaded` implementation.
@rlubos, Do you see any alternative like creating a new `wifi_native` interface, which won't touch `net_wifi_mgmt_offload` and its related stuff?

Reply:
Defining an alternative interface structure is also an option, although as far as I understand, it would still need to expose the offloaded mgmt API? Perhaps we should extract it into a separate struct that would be then included by both interface APIs?

Reply:
So, I have added the `#ifdef` again in there. I honest think we can start with this approach and update in the future. This won't allow both OFFLOADED and NON-OFFLOADED usage. Please, take a look @mniestroj @rlubos.

Reply:
ok, I guess we can live with that for the time being.

Reply:
I think that we confuse socket/networking offloading (like send/recv, TCP/IP stack) with WiFi supplicant offloading (like doing scans, connections, managing WPA, etc.) in this thread. What I still don't buy is including `struct ethernet_api` inside `struct net_wifi_mgmt`. What is the rationale behind it? ESP32 is implementing single function from `struct ethernet_api`, which is `.send()`. Such function exists also in `struct ppp_api` or `struct virtual_interface_api`. So why is it Ethernet and not PPP or Virtual that ESP32 WiFi driver implements?

Looking further into the driver, it seems that the reason why Ethernet is used for ESP32 WiFi is because `esp_wifi_internal_tx()` and `esp_wifi_internal_reg_rxcb()` expect Ethernet frames (i.e. payload prepended with DST, SRC, TYPE/LEN fields). Okay then, let that be, that is what it is **for that driver**. But is that enough reason to actually add `struct ethernet_api` into `struct net_wifi_mgmt`? IMO it is not.

I had a look also into nrf700x driver in NCS repository. It also relies on Ethernet frames for RX/TX in https://github.com/nrfconnect/sdk-nrf/blob/main/drivers/wifi/nrf700x/zephyr/src/zephyr_net_if.c. But why is that? Because it tries to pretend Ethernet device (same as ESP32 WiFi), by converting WiFi (802.11) frames into Ethernet frames.

Both ESP32 and nrf700x drivers are operating on Ethernet frames. To me it looks like a workaround, not something that comes from WiFi specification or will be a common thing for WiFi drivers. So instead of having:
```
#ifdef CONFIG_WIFI_OFFLOAD
	struct net_if_api wifi_iface;
#else
	struct ethernet_api wifi_iface;
#endif
```
which suggest that WiFi offloaded driver are a special case, I would propose:
```
#ifdef CONFIG_WIFI_ESP32
	struct ethernet_api wifi_iface;
#else
	struct net_if_api wifi_iface;
#endif
```
which is a workaround for ESP32 driver trying to emulate Ethernet.

Reply:
@rlubos, any comments regarding the above?  
@mniestroj I get your point regarding that it is specific do ESP32 in this case.  
Should we use more generic approach like `#ifdef CONFIG_WIFI_OVER_ETHERNET_IFACE` for that instead of having custom vendor #ifdef within wifi_mgmt sources?

Reply:
>Should we use more generic approach like #ifdef CONFIG_WIFI_OVER_ETHERNET_IFACE for that instead of having custom vendor #ifdef within wifi_mgmt sources?

That sounds good to me! Probably nrf700x would use the same in the early days.

Reply:
@rlubos, would you review again?

Reply:
Added the `CONFIG_WIFI_OVER_ETHERNET_IFACE` approach.

Reply:
> Both ESP32 and nrf700x drivers are operating on Ethernet frames. To me it looks like a workaround, not something that comes from WiFi specification or will be a common thing for WiFi drivers.

For FullMAC chips we can pass data frames in any format, its mostly `payload + meta_data`, and the UMAC in the chip uses the meta_data to build 802.11 header a.k.a encapsulation. The general norm here is instead of each vendor defining their own meta_data, Ethernet header is used as meta data as it has all necessary information for the UMAC to build 802.11 header. So, even in Linux all FullMAC Wi-Fi chipsets are essentially treated as Ethernet devices in data path.

Wi-Fi chips in Zephyr are doing below, this is not a hack, we definitely need an option for this either default or using `CONFIG_WIFI_OVER_ETHERNET_IFACE` to save some memory.

* esp32 - (FullMAC + 802.1x supplicant)
    - Ethernet APIs + Kconfig based Wi-Fi configuration
* All others (FullMAC + 802.1x supplicant + TCP/IP)
    - wifi_mgmt + WIFI_OFFLOAD (NET_OFFLOAD a.k.a TCP/IP stack offload)
* nrf700x - (only FullMAC)
    -  wifi_mgmt + Native TCP/IP

Reply:
Thanks @krish2718 for your input!

>So, even in Linux all FullMAC Wi-Fi chipsets are essentially treated as Ethernet devices in data path.

Just curious, could you point to some example of that? I was taking a quick look myself, but couldn't find anything useful in Linux tree that we can follow. I just don't have much WiFi experience, so that might be the case.